### PR TITLE
Support complex patterns in DAML REPL

### PR DESCRIPTION
**Pull request by cocreature**
Previously, we only supported simple variable patterns in DAML
repl. This PR extends this to support pretty much all patterns. The
main complexity here is in handling shadowing correctly but I’m
reasonably confident the solution here is sensible.

What doesn’t work properly yet is partial patterns. I’ve added a
comment describing the issue and a potential solution. Given that this
isn’t a regression apart from maybe a slightly worse error message, I
think it makes sense to merge this PR and fix that separately.

changelog_begin

- [DAML REPL - Experimental] You can now use more complex patterns in
  statements, e.g., ``(x,y) <- pure (1,2)``.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/daml-repl/index.rst [hunk 1]
```diff
@@ -60,11 +60,16 @@ two forms:
 1. An expression ``expr`` of type ``Script a`` for some type ``a``. This
    will execute the script ignoring the result.
 
-2. A binding of the form ``x <- expr`` where ``x`` is a variable name
-   and ``expr`` is an expression of type ``Script a``. This will
-   execute the script and bind the result to the variable ``x``. You
+2. A binding of the form ``pat <- expr`` where ``pat`` is pattern, e.g.,
+   a variable name to bind the result to
+   and ``expr`` is an expression of type ``Script a``.
+   This will execute the script and bind the result to the variable ``x``. You
```
Comment (aherrmann-da) [inline docs/source/daml-repl/index.rst]: `x` is no longer defined in the beginning of the sentence.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 2]
```diff
@@ -63,15 +63,51 @@ renderError dflags err = case err of
...
+    go (ListPat ext pats) = ListPat ext (map go pats)
+    go (TuplePat ext pats boxity) = TuplePat ext (map go pats) boxity
+    go (SumPat ext pat tag arity) = SumPat ext (go pat) tag arity
+    go (ConPatIn ext ps) = ConPatIn ext (shadowDetails ps)
+    go ConPatOut{} = error "ConPatOut is never produced by the parser"
+    go p@LitPat{} = p
+    go p@NPat{} = p
+    go NPlusKPat{} = error "N+k patterns are stupid"
+    go (SigPat ext pat sig) = SigPat ext (go pat) sig
+    go SplicePat {} = error "Template haskell is stupid"
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs]: Fair, fixed. Although I don’t expect anyone to actually hit  this.

File: docs/source/daml-repl/index.rst (hunk 1 above)
Comment (cocreature) [reply docs/source/daml-repl/index.rst]: Good catch, fixed!

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 3]
```diff
@@ -63,15 +63,51 @@ renderError dflags err = case err of
...
+  = go (unLoc p)
+  where
+    go p@(VarPat _ var)
+      | occName (unLoc var) `elemOccSet` vars = WildPat noExt
+      | otherwise = p
+    go p@(WildPat _) = p
+    go (LazyPat ext pat) = LazyPat ext (go pat)
+    go (BangPat ext pat) = BangPat ext (go pat)
+    go (AsPat ext a pat)
+        | occName (unLoc a) `elemOccSet` vars = pat
```
[the remark this answers is not in the document]
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs]: Very good catch, I completely missed this!

**Pull request merged**